vendor/knative.dev/hack/codegen-library.sh

```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
+
+# Copyright 2020 The Knative Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Setup the env for doing Knative style codegen.
+
+kn_hack_library=${kn_hack_library:-"$(dirname $0)/../vendor/knative.dev/hack/library.sh"}
```

Comment:
`$(dirname 0)` will be `~/go/src/knative/discovery/vendor/knative.dev/hack` so this'll resolve to `~/go/src/knative/discovery/vendor/knative.dev/vendor/knative.dev/hack/library.sh` which I don't think is what you intend

Reply:
I think `$0` is the file used as the entry point, and this file is intended to be sourced from the current codegen files. Is that not how it works?

Reply:
This works locally and on github actions

Reply:
it depends on how the entrypoint is invoked.  if it's invoked with a relative path, then it'll be relative.  if it's invoked with an absolute path, then it'll be absolute.

BONUS: if it's invoked via a symlink, it can be literally anything you want!  
TRIVIA: this is how busybox compresses dozens of commands into a single binary

Reply:
The main point is `codegen-library.sh` is not invoked, it is sourced.

Reply:
Perhaps Cory's point is that you don't know from where 😉

Reply:
Actually, no.  I think his comment is incorrect, but this does highlight another issue, which is that you are assuming the structure of the repo that's including things. 🤷

Reply:
I am making the assumption that 1- they are using go mod, 2 - they are invoking from a $PROJECT_ROOT/<FOLDER>, and 3- they are vendoring.

Reply:
it is not called the hack repo for nothing